solana/bridge_token_factory/programs/bridge_token_factory/src/state/used_nonces.rs

```diff
@@ -0,0 +1,114 @@
+#[cfg(not(feature = "idl-build"))]
+use bitvec::array::BitArray;
+
+use crate::constants::AUTHORITY_SEED;
+use crate::constants::{USED_NONCES_ACCOUNT_SIZE, USED_NONCES_PER_ACCOUNT};
+#[cfg(not(feature = "idl-build"))]
+use crate::error::ErrorCode;
+
+use super::config::Config;
+
+#[cfg(not(feature = "idl-build"))]
+#[account(zero_copy(unsafe))]
+#[repr(C)]
+pub struct UsedNonces {
+    used: BitArray<[u8; (USED_NONCES_PER_ACCOUNT as usize + 7) / 8]>,
+}
+
+#[cfg(feature = "idl-build")]
+#[account(zero_copy(unsafe))]
+#[repr(C)]
+pub struct UsedNonces {}
+
+impl UsedNonces {
+    pub fn full_rent(rent: &Rent) -> u64 {
+        rent.minimum_balance(USED_NONCES_ACCOUNT_SIZE as usize)
+    }
+
+    pub fn rent_level(nonce: u128, rent: &Rent) -> Result<u64> {
+        let full = Self::full_rent(rent);
+        Ok(
+            ((nonce % USED_NONCES_PER_ACCOUNT as u128 + 1) * full as u128
+                / USED_NONCES_PER_ACCOUNT as u128)
+                .try_into()?,
+        )
+    }
+
+    pub fn use_nonce<'info>(
+        nonce: u128,
+        loader: &AccountLoader<'info, UsedNonces>,
+        config: &mut Account<'info, Config>,
+        rent_reserve: AccountInfo<'info>,
+        payer: AccountInfo<'info>,
+        rent: &Rent,
+        system_program: AccountInfo<'info>,
+    ) -> Result<()> {
+        if config.max_used_nonce < nonce {
+            config.max_used_nonce = nonce;
+        }
+        // use max_used_nonce instead of the requested one to ignore the usage of the nonces from the gap
+        let expected_rent_reserve_lamports =
+            rent.minimum_balance(0) + Self::rent_level(config.max_used_nonce, rent)?;
+        let current_rent_reserve_lamports = rent_reserve.lamports();
+        if current_rent_reserve_lamports < expected_rent_reserve_lamports {
+            // pay for the rent of the next account
+            transfer(
+                CpiContext::new(
+                    system_program,
+                    Transfer {
+                        from: payer,
+                        to: rent_reserve,
+                    },
+                ),
+                expected_rent_reserve_lamports - current_rent_reserve_lamports,
+            )?;
+        } else {
+            // compensate for the account creation
+            let compensation = current_rent_reserve_lamports - expected_rent_reserve_lamports;
+            if compensation > 0 {
+                // compensate expenses for the account creation
+                transfer(
+                    CpiContext::new_with_signer(
+                        system_program,
+                        Transfer {
+                            from: rent_reserve,
+                            to: payer,
+                        },
+                        &[&[AUTHORITY_SEED, &[config.bumps.authority]]],
+                    ),
+                    compensation,
+                )?;
+            }
+        }
+        #[cfg(not(feature = "idl-build"))]
```

Comment:
What is the propose of this feature?

Reply:
@aankor

Reply:
It is turned on for the second compilation used to generate IDL